src/Data/Parameterized/TH/GADT.hs

```diff
@@ -133,10 +135,72 @@ typeVars :: TypeSubstitution a => a -> Set Name
 typeVars = Set.fromList . freeVariables
 
 
--- | @structuralEquality@ declares a structural equality predicate.
+-- | @structuralEquality@ declares a structural equality predicate for a GADT.
 structuralEquality :: TypeQ -> [(TypePat,ExpQ)] -> ExpQ
-structuralEquality tpq pats =
-  [| \x y -> isJust ($(structuralTypeEquality tpq pats) x y) |]
+structuralEquality tpq pats = do
+  d <- reifyDatatype =<< asTypeCon "structuralEquality" =<< tpq
+
+  -- tpq is some type of GADT: data X p1 p2 ... where ...
+  --
+  -- The general approach is to generate a structural type equality such that the
+  -- result is a Maybe (e :+: f) is Just Refl and then verify it is a Just value
+  -- to assert equality by generating (via template haskell):
+  --
+  --   \ x y -> isJust $(structuralTypeEquality ... x y)
+  --
+  -- However, that result presumes a `TestEquality f where testEquality :: f a ->
+  -- f b -> Maybe (a :~: b)`.  If the GADT has a single type parameter, those
```

Comment:
It looks to me like `structuralEquality` calls `structuralTypeEquality`, which generates (the syntax of) a function of type `f a -> f b -> Maybe (a :~: b)`, meaning that using `structuralEquality` would *not* require that `tpq` have an instance of `TestEquality`.